File: app/src/main/java/com/blaxsoftware/directcallwidget/SettingsActivity.java
/*
 * Direct Call Widget - The widget that makes contacts accessible
 * Copyright (C) 2020 Fer P. A.
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.blaxsoftware.directcallwidget;

import android.content.Intent;
import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;

import com.google.android.gms.oss.licenses.OssLicensesMenuActivity;
import com.google.firebase.analytics.FirebaseAnalytics;

import javax.inject.Inject;

import dagger.hilt.android.AndroidEntryPoint;
import dev.ferp.dcw.feature.settings.SettingsFragment;

@AndroidEntryPoint
public class SettingsActivity
        extends AppCompatActivity
        implements SettingsFragment.Callback {

    @Inject
    FirebaseAnalytics mFirebaseAnalytics;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle params = new Bundle();
        params.putString(FirebaseAnalytics.Param.SCREEN_NAME, "Settings");
        params.putString(FirebaseAnalytics.Param.SCREEN_CLASS, SettingsActivity.class.getName());
        mFirebaseAnalytics.logEvent(FirebaseAnalytics.Event.SCREEN_VIEW, params);
        setContentView(R.layout.activity_settings);
        setSupportActionBar(findViewById(R.id.topAppBar));

        final SettingsFragment settingsFragment = new SettingsFragment();
        final StringBuilder versionBuilder = new StringBuilder(BuildConfig.VERSION_NAME);
        if (BuildConfig.DEBUG) {
            versionBuilder.append(" (Debug)");
        }
        Bundle args = new Bundle();
        args.putString(SettingsFragment.Arguments.VERSION_KEY, versionBuilder.toString());
        settingsFragment.setArguments(args);

        getSupportFragmentManager()
                .beginTransaction()
                .replace(R.id.fragContainer, settingsFragment)
                .commit();
    }

    @Override
    public void onOssLicensesClicked() {
        startActivity(new Intent(this, OssLicensesMenuActivity.class));
    }
}

File: app/src/main/java/com/blaxsoftware/directcallwidget/SettingsActivity.kt
package com.blaxsoftware.directcallwidget

import android.appwidget.AppWidgetManager
import android.content.Intent
import android.content.SharedPreferences
import android.os.Bundle
import android.preference.ListPreference
import android.preference.Preference
import android.preference.PreferenceFragment
import android.preference.PreferenceGroup
import android.support.v4.app.NavUtils
import android.support.v7.app.AppCompatActivity
import android.view.MenuItem

import com.blaxsoftware.directcallwidget.R
import com.blaxsoftware.directcallwidget.appwidget.DirectCallWidgetProvider

class SettingsActivity : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val updateWidget = Intent(this, DirectCallWidgetProvider::class.java)
        updateWidget.action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
        sendBroadcast(updateWidget)

        supportActionBar!!.setTitle(R.string.settings)
        fragmentManager.beginTransaction()
                .replace(android.R.id.content, SettingsFragment())
                .commit()
    }

    class SettingsFragment : PreferenceFragment(), SharedPreferences.OnSharedPreferenceChangeListener {

        override fun onCreate(savedInstanceState: Bundle?) {
            super.onCreate(savedInstanceState)
            addPreferencesFromResource(R.xml.preferences)
            updatePreferenceSummaries(preferenceScreen)
        }

        override fun onResume() {
            super.onResume()
            preferenceScreen.sharedPreferences
                    .registerOnSharedPreferenceChangeListener(this)
        }

        override fun onPause() {
            super.onPause()
            preferenceManager.sharedPreferences
                    .unregisterOnSharedPreferenceChangeListener(this)
        }

        override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
            updatePreferenceSummaries(preferenceScreen)
        }

        private fun updatePreferenceSummaries(rootPreference: Preference) {
            if (rootPreference is PreferenceGroup) {
                val pg = rootPreference
                for (i in 0..pg.preferenceCount - 1) {
                    updatePreferenceSummaries(pg.getPreference(i))
                }
            } else if (rootPreference is ListPreference) {
                val lp = rootPreference
                lp.summary = lp.entry
            }
        }
    }
}
